rest_framework_reactive/observer.py

```diff
@@ -127,37 +129,49 @@ def evaluate(self, return_full=True, return_emitted=False):
+                        # Obtain the observer state and lock it. This prevents an observer from being
+                        # processed in parallel from multiple different workers.
+                        observer, _ = models.Observer.objects.select_for_update().get_or_create(
+                            id=self.id,
+                            defaults={
+                                'request': pickle.dumps(self._request),
+                            },
+                        )
+
+                        # Evaluate the observer.
+                        start = time.time()
+                        result = self._evaluate(observer, return_full, return_emitted)
+                        duration = time.time() - start
+
+                        # Log slow observers.
+                        if duration > settings['warnings']['max_processing_time']:
+                            logger.warning(
+                                "Slow observed viewset ({})".format(self._get_logging_id()),
+                                extra=self._get_logging_extra(duration=duration)
+                            )
 
-                # Evaluate the observer.
-                start = time.time()
-                result = self._evaluate(observer, return_full, return_emitted)
-                duration = time.time() - start
+                        # Stop really slow observers.
+                        if duration > settings['errors']['max_processing_time']:
+                            logger.error(
+                                "Stopped extremely slow observed viewset ({})".format(self._get_logging_id()),
+                                extra=self._get_logging_extra(stopped=True, duration=duration)
+                            )
+                            observer.delete()
 
-                # Log slow observers.
-                if duration > settings['warnings']['max_processing_time']:
-                    logger.warning(
-                        "Slow observed viewset ({})".format(self._get_logging_id()),
-                        extra=self._get_logging_extra(duration=duration)
-                    )
+                        return result
+                except IntegrityError:
+                    # If an IntegrityError occurrs we need to rollback the transaction and retry observer
+                    # evaluation as another transaction may be creating an observer concurrently.
+                    if retry == MAX_INTEGRITY_ERROR_RETRIES - 1:
+                        raise
 
-                # Stop really slow observers.
-                if duration > settings['errors']['max_processing_time']:
-                    logger.error(
-                        "Stopped extremely slow observed viewset ({})".format(self._get_logging_id()),
-                        extra=self._get_logging_extra(stopped=True, duration=duration)
-                    )
-                    observer.delete()
+                    continue
```

Comment:
This is unnecessary.

Reply:
Sure but it is more explicit that you want the loop to continue.